### PR TITLE
Add ruff to .pre-commit-config.yaml

### PR DESCRIPTION
**Pull request by tkoyama010**
### Overview

My friend @jrfk told me about [ruff](https://github.com/charliermarsh/ruff). It seems to be increasingly adopted in large projects such as Scipy and Sphinx. (See [Here](https://github.com/charliermarsh/ruff#whos-using-ruff))
Let's try it in PyVista too!

 

### TODO

- [ ] Fix the errors of ruff. 
   - [x] Using [flake8-to-ruff](https://pypi.org/project/flake8-to-ruff/).)
   - [ ] Check test codes have compatibility. 
   - [ ] Wait `inplace=True` deprecate PR #4161
- [x] Discuss removing the above tools.
- [x] Define explicitly which rule to use.

### Details

> Ruff can be used to replace [Flake8](https://pypi.org/project/flake8/) (plus dozens of plugins), [isort](https://pypi.org/project/isort/), [pydocstyle](https://pypi.org/project/pydocstyle/), [yesqa](https://github.com/asottile/yesqa), [eradicate](https://pypi.org/project/eradicate/), [pyupgrade](https://pypi.org/project/pyupgrade/), and [autoflake](https://pypi.org/project/autoflake/), all while executing tens or hundreds of times faster than any individual tool.

- [Python tooling could be much, much faster](https://notes.crmarsh.com/python-tooling-could-be-much-much-faster)
- [Ruff: The First 200 Releases](https://notes.crmarsh.com/ruff-the-first-200-releases)

### CODE REVIEW

**Comment by akaszynski** [comment]
@jorgepiloto has also recommended ruff. Glad to see it implemented here!

**Comment by banesullivan** [comment]
Well, this is neat! Is this ready to go, @tkoyama010?

File: pyproject.toml [hunk 1]
```diff
@@ -165,3 +165,29 @@ image_cache_dir = "tests/plotting/image_cache"
 
 [tool.mypy]
 ignore_missing_imports = true
+
+[tool.ruff]
+exclude = [
```
Comment (jorgepiloto) [inline pyproject.toml:170]: I would suggest also to exclude a possible `venv` and `.venv` directories, @tkoyama010 @banesullivan.

**Comment by tkoyama010** [comment]
@banesullivan Thank you. I have read the documentation carefully and it seems it needs to be configured to enable each rule. Please give me some time to investigate.

File: examples/02-plot/vertices.py [hunk 2]
```diff
@@ -1,10 +1,8 @@
-"""
-Visible Vertices
-~~~~~~~~~~~~~~~~
+"""Visible Vertices
+~~~~~~~~~~~~~~~~.
```
Comment (akaszynski) [inline examples/02-plot/vertices.py:2]: Seems that it doesn't do a great job with titles for these examples. We can "cheat" by providing a reference at the top. Probably a good idea anyway.

```suggestion
""".. _visible_vertices_example:

Visible Vertices
~~~~~~~~~~~~~~~~.

```

File: examples/02-plot/vertices.py [hunk 3]
```diff
@@ -1,10 +1,8 @@
...
 
 Display vertices on a mesh in the same fashion as edge visibility.
 """
-# sphinx_gallery_thumbnail_number = 3
```
Comment (akaszynski) [inline examples/02-plot/vertices.py:7]: Was this intentional?

File: tests/utilities/test_reader.py [hunk 4]
```diff
@@ -528,7 +528,7 @@ def test_pvdreader_no_part_group():
 
     reader.set_active_time_value(1.0)
     assert len(reader.active_datasets) == 2
-    for i, dataset in enumerate(reader.active_datasets):
+    for _i, dataset in enumerate(reader.active_datasets):
```
Comment (akaszynski) [inline tests/utilities/test_reader.py:531]: I think using `ii` would be better here (and elsewhere).

File: examples/02-plot/clear.py [hunk 5]
```diff
@@ -1,12 +1,10 @@
-"""
-Clearing a Mesh or the Entire Plot
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""Clearing a Mesh or the Entire Plot
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/02-plot/clear.py:2]:
```suggestion
""".. _clear_example:
Clearing a Mesh or the Entire Plot
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/02-plot/ghost-cells.py [hunk 6]
```diff
@@ -1,6 +1,5 @@
-"""
-Hide Cells with Ghosting
-~~~~~~~~~~~~~~~~~~~~~~~~
+"""Hide Cells with Ghosting
+~~~~~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/02-plot/ghost-cells.py:2]:
```suggestion
""".. _ghost_cells_example:
Hide Cells with Ghosting
~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/02-plot/image_depth.py [hunk 7]
```diff
@@ -1,13 +1,11 @@
-"""
-Render a depth image
-~~~~~~~~~~~~~~~~~~~~
+"""Render a depth image
+~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/02-plot/image_depth.py:2]:
```suggestion
""".. _image_depth_example
Render a depth image
~~~~~~~~~~~~~~~~~~~~
```

File: examples/02-plot/labels.py [hunk 8]
```diff
@@ -1,14 +1,11 @@
-"""
-Label Points
-~~~~~~~~~~~~
+"""Label Points
+~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/02-plot/labels.py:2]:
```suggestion
""".. _labels_example:
Label Points
~~~~~~~~~~~~
```

File: examples/02-plot/multi-window.py [hunk 9]
```diff
@@ -1,6 +1,5 @@
-"""
-Multi-Window Plot
-~~~~~~~~~~~~~~~~~
+"""Multi-Window Plot
+~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/02-plot/multi-window.py:2]:
```suggestion
""".. _multi_window_example
Multi-Window Plot
~~~~~~~~~~~~~~~~~
```

File: examples/02-plot/vertices.py (hunk 2 above)
Comment (tkoyama010) [reply examples/02-plot/vertices.py:2]:
```suggestion
""".. _vertices_example:
Visible Vertices
~~~~~~~~~~~~~~~~
```

File: examples/00-load/read-dolfin.py [hunk 10]
```diff
@@ -1,6 +1,5 @@
-"""
-Read FEniCS/Dolfin Meshes
-~~~~~~~~~~~~~~~~~~~~~~~~~
+"""Read FEniCS/Dolfin Meshes
+~~~~~~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/00-load/read-dolfin.py:2]:
```suggestion
""".. _read_dolfin_example:
Read FEniCS/Dolfin Meshes
~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/00-load/read-image.py [hunk 11]
```diff
@@ -1,6 +1,5 @@
-"""
-Read Image Files
-~~~~~~~~~~~~~~~~
+"""Read Image Files
+~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/00-load/read-image.py:2]:
```suggestion
""".. _read_image_example:
Read Image Files
~~~~~~~~~~~~~~~~
```

File: examples/01-filter/distance-between-surfaces.py [hunk 12]
```diff
@@ -1,6 +1,5 @@
-"""
-Distance Between Two Surfaces
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""Distance Between Two Surfaces
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/01-filter/distance-between-surfaces.py:2]:
```suggestion
""".. _distance_between_surfaces:
Distance Between Two Surfaces
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/01-filter/extract-cells-inside-surface.py [hunk 13]
```diff
@@ -1,22 +1,20 @@
-"""
-Extract Cells Inside Surface
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""Extract Cells Inside Surface
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/01-filter/extract-cells-inside-surface.py:2]:
```suggestion
""".. _extract_cells_inside_surface_example:
Extract Cells Inside Surface
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/01-filter/extrude-rotate.py [hunk 14]
```diff
@@ -1,5 +1,4 @@
-"""
-Extrude Rotation
+"""Extrude Rotation
```
Comment (tkoyama010) [inline examples/01-filter/extrude-rotate.py:1]:
```suggestion
""".. _extrude_rotate_example:
Extrude Rotation
```

File: examples/01-filter/project-plane.py [hunk 15]
```diff
@@ -1,12 +1,10 @@
-"""
-Project to a Plane
-~~~~~~~~~~~~~~~~~~
+"""Project to a Plane
+~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/01-filter/project-plane.py:2]:
```suggestion
""".. _project_plane_example:
Project to a Plane
~~~~~~~~~~~~~~~~~~
```

File: examples/01-filter/sampling_functions_3d.py [hunk 16]
```diff
@@ -1,5 +1,4 @@
-"""
-Sample Function: Perlin Noise in 3D
+"""Sample Function: Perlin Noise in 3D
```
Comment (tkoyama010) [inline examples/01-filter/sampling_functions_3d.py:1]:
```suggestion
""".. _sampling_functions_3d_example:
Sample Function: Perlin Noise in 3D
```

File: examples/02-plot/plot-over-circular-arc.py [hunk 17]
```diff
@@ -1,12 +1,10 @@
-"""
-Plot Scalars Over a Circular Arc
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""Plot Scalars Over a Circular Arc
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/02-plot/plot-over-circular-arc.py:2]:
```suggestion
""".. _plot_over_circular_arc_example:
Plot Scalars Over a Circular Arc
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/02-plot/screenshot.py [hunk 18]
```diff
@@ -1,10 +1,8 @@
-"""
-Saving Screenshots
-~~~~~~~~~~~~~~~~~~
+"""Saving Screenshots
+~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/02-plot/screenshot.py:2]:
```suggestion
""".. _screenshot_example:
Saving Screenshots
~~~~~~~~~~~~~~~~~~
```

File: examples/02-plot/spherical.py [hunk 19]
```diff
@@ -1,6 +1,5 @@
-"""
-Plot data in spherical coordinates
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""Plot data in spherical coordinates
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/02-plot/spherical.py:2]:
```suggestion
""".. _spherical_example:
Plot data in spherical coordinates
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/02-plot/vector-component.py [hunk 20]
```diff
@@ -1,6 +1,5 @@
-"""
-Plot Vector Component
-~~~~~~~~~~~~~~~~~~~~~
+"""Plot Vector Component
+~~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/02-plot/vector-component.py:2]:
```suggestion
""".. _vector_component_example:
Plot Vector Component
~~~~~~~~~~~~~~~~~~~~~
```

File: examples/03-widgets/multi-slider-widget.py [hunk 21]
```diff
@@ -1,6 +1,5 @@
-"""
-Multiple Slider Widgets
-~~~~~~~~~~~~~~~~~~~~~~~
+"""Multiple Slider Widgets
+~~~~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/03-widgets/multi-slider-widget.py:2]:
```suggestion
""".. _multi_slider_widget_example:
Multiple Slider Widgets
~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/03-widgets/sphere-widget.py [hunk 22]
```diff
@@ -1,6 +1,5 @@
-"""
-Sphere Widget
-~~~~~~~~~~~~~
+"""Sphere Widget
+~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/03-widgets/sphere-widget.py:2]:
```suggestion
""".. _sphere_widget_example:
Sphere Widget
~~~~~~~~~~~~~
```

File: examples/03-widgets/spline-widget.py [hunk 23]
```diff
@@ -1,6 +1,5 @@
-"""
-Spline Widget
-~~~~~~~~~~~~~
+"""Spline Widget
+~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/03-widgets/spline-widget.py:2]:
```suggestion
""".. _spline_widget_example:
Spline Widget
~~~~~~~~~~~~~
```

File: examples/99-advanced/plotting-algorithms.py [hunk 24]
```diff
@@ -1,6 +1,5 @@
-"""
-Plotting with VTK Algorithms
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""Plotting with VTK Algorithms
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/99-advanced/plotting-algorithms.py:2]:
```suggestion
""".. _plotting_algorithms_example:
Plotting with VTK Algorithms
~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: examples/99-advanced/ray-trace.py [hunk 25]
```diff
@@ -1,6 +1,5 @@
-"""
-Ray Tracing
-~~~~~~~~~~~
+"""Ray Tracing
+~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/99-advanced/ray-trace.py:2]:
```suggestion
""".. _ray_trace_example:

Ray Tracing
~~~~~~~~~~~
```

File: examples/99-advanced/warp-by-vector-eigenmodes.py [hunk 26]
```diff
@@ -1,6 +1,5 @@
-"""
-Displaying eigenmodes of vibration using ``warp_by_vector``
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""Displaying eigenmodes of vibration using ``warp_by_vector``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline examples/99-advanced/warp-by-vector-eigenmodes.py:2]:
```suggestion
""".. _warp_by_vector_eigenmodes_example:

Displaying eigenmodes of vibration using ``warp_by_vector``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: tests/utilities/test_reader.py (hunk 4 above)
Comment (tkoyama010) [reply tests/utilities/test_reader.py:532]: I guess we can remove `i` here.
```suggestion
    for dataset in reader.active_datasets:
```

File: pyvista/ext/coverage.py [hunk 27]
```diff
@@ -1,23 +1,22 @@
-"""
-    sphinx.ext.coverage
-    ~~~~~~~~~~~~~~~~~~~
+"""sphinx.ext.coverage
+~~~~~~~~~~~~~~~~~~~.
```
Comment (tkoyama010) [inline pyvista/ext/coverage.py:2]:
```suggestion
""".. _ref_sphinx_ext_coverage:
sphinx.ext.coverage
~~~~~~~~~~~~~~~~~~~
```

File: pyvista/core/datasetattributes.py [hunk 28]
```diff
@@ -246,7 +249,7 @@ def __delitem__(self, key: str):
 
     def __contains__(self, name: str) -> bool:
         """Implement the ``in`` operator."""
-        return name in self.keys()
+        return name in self
```
Comment (tkoyama010) [inline pyvista/core/datasetattributes.py:252]: This is from `SIM118` in [`flake8-simplify-sim`](https://beta.ruff.rs/docs/rules/#flake8-simplify-sim).
This class is not a dictionary and does not work well. I would cheat use the following expression.
```suggestion
        keys = self.keys()
        return name in keys
```

File: pyvista/core/datasetattributes.py [hunk 29]
```diff
@@ -972,7 +989,7 @@ def values(self) -> List[pyvista_ndarray]:
         [pyvista_ndarray([0, 0, 0, 0, 0, 0]), pyvista_ndarray([0, 1, 2, 3, 4, 5])]
 
         """
-        return [self.get_array(name) for name in self.keys()]
+        return [self.get_array(name) for name in self]
```
Comment (tkoyama010) [inline pyvista/core/datasetattributes.py:992]:
```suggestion
        keys = self.keys()
        return [self.get_array(name) for name in keys]
```

File: tests/test_composite.py [hunk 30]
```diff
@@ -707,7 +717,7 @@ def test_set_active_scalars_components(multiblock_poly):
 
     multiblock_poly.set_active_scalars(None)
     multiblock_poly.set_active_scalars('data')
-    for block in multiblock_poly:
+    for _ in multiblock_poly:
         assert multiblock_poly[0].point_data.active_scalars_name == 'data'
```
Comment (tkoyama010) [inline tests/test_composite.py:721]: We need to check what this line want to do by loop.

File: doc/source/make_external_gallery.py [hunk 31]
```diff
@@ -29,169 +29,164 @@ def format(self):
...
     ),
-    magpylib=Example(
+    "magpylib": Example(
         title="Coil Field Lines example in Magpylib",
         description="Pyvista streamlines of Coil Field Lines",
         link="https://magpylib.readthedocs.io/en/latest/examples/examples_30_coil_field_lines.html#pyvista-streamlines",
         image="coil_field_lines.png",
     ),
-    pyfbs=Example(
+    pyfbs: Example(
```
Comment (tkoyama010) [inline doc/source/make_external_gallery.py:183]:
```suggestion
    "pyfbs": Example(
```

File: pyvista/ext/coverage.py [hunk 32]
```diff
@@ -1,23 +1,23 @@
-"""
-    sphinx.ext.coverage
-    ~~~~~~~~~~~~~~~~~~~
+""".. _ref_sphinx_ext_coverage:
+sphinx.ext.coverage
+~~~~~~~~~~~~~~~~~~~
```
Comment (tkoyama010) [inline pyvista/ext/coverage.py:3]:
```suggestion

sphinx.ext.coverage
~~~~~~~~~~~~~~~~~~~
```

**Comment by tkoyama010** [comment]
This pull request is too huge. I will split this pull request into several parts and ask you to review them there. (I will keep this pull request open as a final destination.)

**Comment by banesullivan** [comment]
I'm concerned about the level of code changes happening in this PR. From what I understand, ruff is simply a mechanism under which the same checks/rules are enforced and thus the rule should remain the same and there should not be any code  or formatting changes. @tkoyama010, would you please clarify why there are so many actual changes happening here?

**Comment by tkoyama010** [comment]
@banesullivan Yes, I am splitting the pull request to isolate that very reason. Thanks for comment.

**Comment by tkoyama010** [comment]
Follow #4170 closing.

**Pull request closed without merging**